design/003-service-topology-support.md

```diff
@@ -0,0 +1,118 @@
+    2. In the proxy, get the value from the environment variable.
+
+* **Endpoint weights**: in `api-resolve` crate change the metadata to support endpoint topology labels. I propose the topology labels are either set as an explicit, separate value, or be included as labels as part of the existing `labels` map.  
+
+* **Other changes**: add support for the gRPC call mentioned in the summary, to retrieve topology preference associated with a service and topology labels of the node the proxy is running on.
+
+### Destination Service
+---
+
+The destination service needs to support the new `endpointSlices` feature, the `service topology` feature and make changes to the `gRPC` api server to support the new proxy call.
+
+* **Endpoint slices**: to implement support for this, there should be a new watcher in place to register all `endpointSlices` resources. In practice, there might be some complications. The destination api should use either endpoint slices or endpoints, both at the same time would result in useless calls and duplicate information (I assume). I will discuss this a bit more in depth in the [constraints subsection](#constraints). 
+    - Each endpoint slice can hold 100 endpoints for a given service. Also, each endpoint slice has a generated name based on the service name. To construct the service id in the respective handler function we need to make use of the `service-name` annotation of the endpoint slice rather than its name directly (otherwise each service would be created more than once but under different names).
+    - Every address in the address set should now retain topology metadata. Since the order is not specific, it can be incorporated in the labels map that is already part of the struct.
+
+* **Service topology**: extend the current `servicePublisher` implementation to include the topology preference as described in the service spec. This implies making changes to the handler functions of the watcher.
+
+
+* **Adding weights to addresses**: to add weights to an address, we need access to the service topology, the source node (its topology labels) and the endpoint topology. If the source node name is sent by the proxy, the node's topology values can be trivially accessed using the `K8s API`. There are two edge cases that exist under my assumptions however, which makes weighting the addresses a bit harder than it should be on the destination service side. First, two source proxies can listen on the same service, because of this, I think it is infeasible to add the weighting logic on subscription. Second, we have to also consider updates -- while on update we can weight a endpoint based on the current endpoints that already exist in a stream, there is a high chance there might be no available endpoints, so inferring weights from existing endpoints when updating also does not sound feasible (even if there is a 1:1 relation between a source and a service stream).
```

Comment:
I'm not sure I totally understand your points here about not being able to make topology decisions on subscription or update.

Here's my first instinct for how I would approach this:

If we add a `nodeName` context token field to the Destination.Get method, then the destination controller should have all the information it needs to make topology aware routing decisions and return the correct list of endpoints to the proxy (i.e. the endpoints with the highest priority for that node).

I think this decision can be made in the EndpointTraslator.  If the EndpointTraslator has the list of endpoints (including their topology metadata), the topology keys from the service, and the node name of the subscribing proxy, it can look up the topology labels for that node and use that to filter down the list of endpoints to only the ones that match the node's label for the first topology key.

Reply:
I think this approach simplifies a lot of things:
* the only proxy change is that it needs to send it's node name (discovered via the downward API and an env variable).  the proxy does not need any topology aware routing logic since the destination controller will just give it the correct set of endpoints to route to.
* no need for a new separate API to discover topology keys

Reply:
@adleong  thank you so much for your comments! Disregard the inability to make topology decisions on subscription or update, that was before I realised `weights` are to be `priorities`. My whole design focused on the ability to assign weights on the fly based on availability and the topological domain; assigning priorities instead makes this redundant and I don't see any reason why it should be done on the proxy side, so thank you for clarifying.

It makes a lot of sense to change the proposal based on your comments and guidelines, and I do agree that it brings a major simplification on the design. I have started to re-work the proposal. There is one thing that I want to address before I push any changes, however.

> filter down the list of endpoints to only the ones that match the node's label for the first topology key

I have thought about this and I realise it might sound a bit nitpicky, but what would happen if by chance all source pods would be placed on the same node whereas the destination service's pods are evenly spaced out in a cluster? I realise this might be a product of randomness and might not happen easily, but it could also be a product of affinity in certain cases.

Going on my previous example, if service A (whose pods are on the same node) calls service B (whose pods are evenly spaced out), then all of the calls would be routed to the same service B pod (?). Would it make sense to account for such cases by sending additional endpoints from the second topology key that matches the node's label?

Reply:
This is a valid concern but I think we should stick to the service topology specification: https://kubernetes.io/docs/concepts/services-networking/service-topology/

> Traffic will be directed to a Node whose value for the first label matches the originating Node’s value for that label. If there is no backend for the Service on a matching Node, then the second label will be considered, and so forth, until no labels remain.

So, the case that you describe, all traffic from A to B should go to the B pod which is on the same node as the A pods.

Reply:
Alright, that makes sense to me.